### PR TITLE
OV-377: Create preview page and flow

### LINKED ISSUE #377: FEAT: Create preview page
### What feature?

- [x] Create a separate Preview page.
- [x] Add a video player there for playback (use the same one as on the home page in the modal).
- [x] Add a button to copy the preview video link (for each card on the home page).
- [x] When the button is clicked, send a request to the backend to generate a JWT with the video ID.
- [x] Add an endpoint to generate a preview URL with a JWT.
- [x] The server response should be a URL string with a JWT and videoId (preview/:jwt).
- [x]  When the response arrives, it should be copied to the clipboard.
- [x]  When a user accesses this URL, the JWT with the videoId should be extracted from the page URL, and a request should be sent to decode the JWT and retrieve the videoId, from which the videoUrl will be obtained.
- [x]  Send the video URL to the frontend.

Note:

- The JWT token should be handled for the URL since the browser address bar should not contain dots.

### Add screenshots

_No response_

### Acceptance criteria

_No response_

### PR DESCRIPTION
**Pull request by stefano-lacorazza**
### Summary:
This pull request implements the following features:

1. **Preview Page Creation**:  
   - Created a whitelisted separate Preview page, accessible via a generated JWT-based URL.
   - Added a video player to this page for playback of the video.

2. **Button for Generating Preview Link**:  
   - Added a button to each rendered video on the homepage to copy the preview link.
   - When the button is clicked, a request is sent to the backend to generate a JWT containing the video ID.

3. **JWT-based Preview URL Generation**:  
   - Added an endpoint to generate a preview URL with a JWT and the videoId.  
   - The server response returns a URL string with the format `preview/:jwt`, which includes the JWT and the video ID.
   - Upon receiving the server response, the preview URL is copied to the clipboard.

4. **Preview URL Access and JWT Decoding**:  
   - When a user accesses the preview URL, the JWT and videoId are extracted from the page URL.
   - A request is sent to decode the JWT and retrieve the videoId.
   - The decoded videoId is then used to fetch the corresponding video URL for playback.

5. **New Public-Video Controller**:  
   - Implemented a new `public-video` controller to handle requests related to video previews, including decoding JWTs and retrieving video information.

https://github.com/user-attachments/assets/9b5c8b7d-5fb2-43fc-9c41-9bb668c8b24b

---

### CODE REVIEW

File: frontend/src/bundles/home/helpers/create-url-token.ts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+
+const createVideoUrl = (jwtToken: string): string => {
+    const baseUrl = 'http://localhost:3000';
```
Comment (stefano-lacorazza) [inline frontend/src/bundles/home/helpers/create-url-token.ts]: Here how could I create the baseUrl dynamically for deployment?

File: backend/src/common/services/token/token.services.ts [hunk 2]
```diff
@@ -17,6 +17,13 @@ class TokenService {
             .sign(this.secretKey);
     }
 
+    public async createVideoIdToken(videoId: string): Promise<string> {
+        const jwt = await new SignJWT({ videoId })
+            .setProtectedHeader({ alg: 'HS256' })
+            .sign(this.secretKey);
+        return jwt.replaceAll('.', '~');
+    }
+
```
Comment (anton-otroshchenko) [inline backend/src/common/services/token/token.services.ts]: lets move it to video service

Comment (anton-otroshchenko) [reply backend/src/common/services/token/token.services.ts]: can we use createToken method and set expiration time optional?

File: backend/src/common/services/token/token.services.ts [hunk 3]
```diff
@@ -30,6 +37,11 @@ class TokenService {
         const payload = await this.verifyToken(token);
         return (payload?.['userId'] as string) || null;
     }
+
+    public async getVideoIdFromToken(token: string): Promise<string | null> {
+        const payload = await this.verifyToken(token);
+        return (payload?.['videoId'] as string) || null;
```
Comment (anton-otroshchenko) [inline backend/src/common/services/token/token.services.ts]:
```suggestion
        return (payload?.['videoId'] as string) ?? null;
```

File: frontend/src/bundles/preview/pages/preview.tsx [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+import styles from './styles.module.css';
+
+type Properties = {
+    jwt: string;
+};
+
+const Preview: React.FC<Properties> = ({ jwt }) => {
+    const dispatch = useAppDispatch();
+    const [url, setUrl] = useState<string>('');
+    const [loading, setLoading] = useState<boolean>(true);
```
Comment (anton-otroshchenko) [inline frontend/src/bundles/preview/pages/preview.tsx]: isLoading, setIsLoading

**Review by anton-otroshchenko — CHANGES_REQUESTED** [review]

File: frontend/src/bundles/home/helpers/create-url-token.ts [hunk 5]
```diff
@@ -0,0 +1,10 @@
+const createVideoUrl = (jwtToken: string): string => {
+    const baseUrl =
+        import.meta.env['VITE_APP_NODE_ENV'] === 'production'
```
Comment (anton-otroshchenko) [inline frontend/src/bundles/home/helpers/create-url-token.ts]: please use our class on frontend to handle .env

Comment (anton-otroshchenko) [reply frontend/src/bundles/home/helpers/create-url-token.ts]: also use this 
image

**Review by anton-otroshchenko — CHANGES_REQUESTED** [review]

File: frontend/src/bundles/home/helpers/create-url-token.ts [hunk 6]
```diff
@@ -0,0 +1,13 @@
+import { AppEnvironment } from '~/bundles/common/enums/enums.js';
+import { config } from '~/framework/config/config.js';
+
+const createVideoUrl = (jwtToken: string): string => {
+    const baseUrl =
+        config.ENV.APP.ENVIRONMENT === AppEnvironment.PRODUCTION
+            ? 'http://bsa-2024-outreachvids-dev.eu-north-1.elasticbeanstalk.com'
+            : 'http://localhost:3000';
+
```
Comment (sofiia-trokhymchuk) [inline frontend/src/bundles/home/helpers/create-url-token.ts]: Add a new env variable and use it. What if the deployment URL changes?

File: backend/src/bundles/public-video/public-video.controller.ts [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    private async findUrlByToken(
+        options: ApiHandlerOptions,
+    ): Promise<ApiHandlerResponse> {
+        const headers = options.headers as Record<string, { value: string }>;
+        const videoTokenHeader = headers['video_token']?.toString() ?? '';
+
+        return {
+            status: HTTPCode.OK,
+            payload: await this.publicVideoService.findUrlByToken(
+                videoTokenHeader ?? '',
```
Comment (sofiia-trokhymchuk) [inline backend/src/bundles/public-video/public-video.controller.ts]: Why do you need this nullish coalescing here? You already check it on line 32

File: backend/src/bundles/public-video/public-video.service.ts [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+            this.throwVideoNotFoundError();
+        }
+
+        const video = await this.videoRepository.findById(id);
+
+        if (!video) {
+            this.throwVideoNotFoundError();
+        }
+
+        const url = video.toObject().url;
```
Comment (sofiia-trokhymchuk) [inline backend/src/bundles/public-video/public-video.service.ts]:
```suggestion
        const { url } = video.toObject();
```

File: backend/src/bundles/videos/video.controller.ts [hunk 9]
```diff
@@ -109,6 +109,17 @@ class VideoController extends BaseController {
                     }>,
                 ),
         });
+        this.addRoute({
+            path: `${VideosApiPath.ID}/share`,
```
Comment (sofiia-trokhymchuk) [inline backend/src/bundles/videos/video.controller.ts]: Add this path to the enum

File: frontend/src/bundles/common/api/video-api/videos-api.ts [hunk 10]
```diff
@@ -81,6 +97,24 @@ class VideosApi extends BaseHttpApi {
 
         await response.json<boolean>();
     }
+
+    public async getVideoIdJWT(id: string): Promise<string> {
+        const response = await this.load(
+            this.getFullEndpoint(`${VideosApiPath.ROOT}${id}/share`, {}),
```
Comment (sofiia-trokhymchuk) [inline frontend/src/bundles/common/api/video-api/videos-api.ts]: Same here

**Review by sofiia-trokhymchuk — CHANGES_REQUESTED** [review]

File: frontend/src/bundles/home/helpers/create-url-token.ts (hunk 6 above)
Comment (stefano-lacorazza) [reply frontend/src/bundles/home/helpers/create-url-token.ts]: Done but could you test it if it was done correctly before merging? I have no way of testing in production enviroment and I dont want to break something a day before the demo. Thank you

File: frontend/src/bundles/home/helpers/create-url-token.ts [hunk 11]
```diff
@@ -4,7 +4,7 @@ import { config } from '~/framework/config/config.js';
 const createVideoUrl = (jwtToken: string): string => {
     const baseUrl =
         config.ENV.APP.ENVIRONMENT === AppEnvironment.PRODUCTION
-            ? 'http://bsa-2024-outreachvids-dev.eu-north-1.elasticbeanstalk.com'
+            ? config.ENV.DEPLOYMENT.URL
```
Comment (sofiia-trokhymchuk) [inline frontend/src/bundles/home/helpers/create-url-token.ts]: It is better to add something like PUBLIC_URL. And use it as it is without checks

File: frontend/.env.example [hunk 12]
```diff
@@ -10,3 +10,9 @@ VITE_APP_API_ORIGIN_URL=/api/v1
 #
 VITE_APP_PROXY_SERVER_URL=http://localhost:3001
 VITE_APP_SOCKET_ORIGIN_URL=/socket.io
+
+#
+# DEPLOYMENT
+#
+DEPLOYMENT_URL=http://bsa-2024-outreachvids-dev.eu-north-1.elasticbeanstalk.com
```
Comment (sofiia-trokhymchuk) [inline frontend/.env.example]:
```suggestion
#
# DEPLOYMENT
#
PUBLIC_URL=http://localhost:3000
```

**Review by sofiia-trokhymchuk — CHANGES_REQUESTED** [review]

File: frontend/src/bundles/home/helpers/create-url-token.ts (hunk 11 above)
Comment (stefano-lacorazza) [reply frontend/src/bundles/home/helpers/create-url-token.ts]: Im having problems with this one. I changed .env,  EnvironmentSchema, and base-config.package adding the new ENV PUBLIC_URL variable but when I try to access it in createVideoUrl I keep getting undefined.

**Review by sofiia-trokhymchuk — APPROVED** [review]

**Review by nikita-remeslov — APPROVED** [review]

**Pull request merged**